crates/staking/src/lib.rs

```diff
@@ -823,3 +832,258 @@ where
+                assert_total_rewards(0);
+            });
+        }
+
+        /// like setup_broken_state, but after depositing such a large sum that you can withdraw
+        /// despite the slash bug (it will withdraw an incorrect but non-zero amount)
+        fn setup_broken_state_with_withdrawable_reward() {
+            use mock::*;
+
+            setup_broken_state();
+            assert_total_rewards(12000);
+            assert_ok!(Staking::distribute_reward(Token(IBTC), &VAULT, fixed!(1_000_000)));
+            assert_total_rewards(1_012_000);
+        }
+
+        #[test]
+        fn test_broken_state_with_withdrawable_amount() {
+            use mock::*;
+            run_test(|| {
+                setup_broken_state_with_withdrawable_reward();
+                assert_total_rewards(1_012_000);
+
+                // check that we can indeed withdraw some non-zero amount
+                assert_ok!(
+                    Staking::withdraw_reward(Token(IBTC), &VAULT, &VAULT.account_id),
+                    967_000
+                );
+                assert_total_rewards(1_012_000 - 967_000);
+
+                assert_ok!(Staking::withdraw_reward(Token(IBTC), &VAULT, &VAULT.account_id), 0);
+            });
+        }
+
+        #[test]
+        fn test_migration_of_account_with_withdrawable_amount() {
+            use mock::*;
+            run_test(|| {
+                let fee_pool_account_id = 23;
+
+                assert_ok!(<orml_tokens::Pallet<Test> as MultiCurrency<
+                    <Test as frame_system::Config>::AccountId,
+                >>::deposit(
+                    Token(IBTC), &fee_pool_account_id, 10_000_000,
+                ));
+
+                setup_broken_state_with_withdrawable_reward();
+                assert_total_rewards(1_012_000);
+
+                // now simulate that we deploy the fix and do the migration
+
+                assert_ok!(fix_broken_state::<Test, _>(fee_pool_account_id));
+                assert_total_rewards(0);
+                assert_eq!(
+                <orml_tokens::Pallet<Test> as MultiCurrency<<Test as frame_system::Config>::AccountId>>::free_balance(
+                    Token(IBTC),
+                    &VAULT.account_id
+                ),
+                1_012_000
+            );
+
+                // check that we can't withdraw any additional amount
+                assert_ok!(Staking::withdraw_reward(Token(IBTC), &VAULT, &VAULT.account_id), 0);
+
+                // check that behavior is back to normal
+                assert_ok!(Staking::distribute_reward(Token(IBTC), &VAULT, fixed!(3000)));
+                assert_total_rewards(3000);
+                assert_ok!(Staking::withdraw_reward(Token(IBTC), &VAULT, &VAULT.account_id), 3000);
+                assert_total_rewards(0);
+            });
+        }
+    }
+
+    #[transactional]
+    pub fn fix_broken_state<T, U>(fee_pool_account_id: T::AccountId) -> DispatchResult
```

Comment:
Not sure if we should test this against Kintsugi data? It shouldn't fail but not sure what the effort would be to verify that.